### PR TITLE
Add /plot endpoint to backoffice

### PR DESCRIPTION
**Pull request by Deathn0t**

### CODE REVIEW

**Comment by rth** [comment]
Thanks a lot @Deathn0t ! As a initial version of the /plot endpoint overall this LGTM, I'll apply a few fixes to make the CI pass and call it in tests. Once it's merged can then iterate from there.

**Comment by olivierteboul** [comment]
Hi Romain, could you move this to the backoffice please ? those are not supposed to be accessible.
ping me if needed

**Comment by rth** [comment]
>  could you move this to the backoffice please ?

On it now.

**Comment by rth** [comment]
I moved /plot to the backoffice. I would like to merge this as an initial version  as is, the actual layout of the page content and the figures can be added/improved in follow up PRs.

We need better (more explicit) names and titles for the /plot and /dashboard endpoints.

File: icubam/backoffice/handlers/plot.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from bokeh.plotting import figure
+from icubam.backoffice.handlers import base
+
+from pprint import pprint
+
+
+class PlotHandler(base.BaseHandler):
+  ROUTE = '/plot'
+
+  def initialize(self, config, db):
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:15]: you can skip this. the initialize from the parent will be use automatically

File: icubam/backoffice/handlers/plot.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      title="Lits Disponibles Par Ville"
+    )
+    p.xaxis.major_label_orientation = math.pi / 2
+    p.y_range.start = 0
+
+    p.vbar(x=cities_names, top=beds_available, width=0.9)
+
+    script, div = components(p)
+    figures.append(dict(script=script, div=div))
+    # for i, el in enumerate(bed_counts):
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:55]: remove comments if not used

File: icubam/backoffice/handlers/plot.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  @tornado.web.authenticated
+  async def get(self):
+    """Serves a page with a table gathering current bedcount data with some extra information."""
+    data = {}
+    figures = []
+
+    # plot
+    # bed_counts = self.db.get_bed_counts()
+    bed_counts = self.db.get_visible_bed_counts_for_user(None, force=True)
+    cities = {}
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:27]: what about a collections.defaultdict(int) ? that would simplify the loop below with a single line
cities[curr.icu.city] += curr.n_covid_free

File: icubam/backoffice/handlers/plot.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    for curr in bed_counts:
+      curr = curr.to_dict()
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:29]: it does not seems like you need a dict

File: icubam/backoffice/handlers/plot.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      city_name = curr['icu']['city']
+      val = cities.get(city_name, 0)
+      val += curr['n_covid_free']
+      cities[city_name] = val
+
+    get_val = lambda tup: tup[1]
+    cities = [(name, val) for name, val in cities.items()]
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:36]: cities = list(cities.items())

File: icubam/backoffice/handlers/plot.py (hunk 3 above)
Comment (rth) [reply icubam/backoffice/handlers/plot.py]: Good point!

**Comment by rth** [comment]
Thanks @olivierteboul. Removed comments, and simplified code somewhat.

File: icubam/backoffice/handlers/plot.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+class PlotHandler(base.BaseHandler):
+  ROUTE = '/plot'
+
+  @tornado.web.authenticated
+  async def get(self):
+    """Serves a page with a table gathering current bedcount data with some extra information."""
+    data = {}
+    figures = []
+
+    bed_counts = self.db.get_visible_bed_counts_for_user(None, force=True)
```
Comment (olivierteboul) [inline icubam/backoffice/handlers/plot.py:22]: could you not force and pass as only argument self.user.user_id ? this way the plot will be customized for the user.

File: icubam/backoffice/server_test.py [hunk 7]
```diff
@@ -42,9 +42,10 @@ def test_homepage_without(self):
...
       dashboard.ListBedCountsHandler, icus.ListICUsHandler,
       users.ListUsersHandler, tokens.ListTokensHandler,
+      plot.PlotHandler
       ]
     for handler in handlers:
       with mock.patch.object(handler, 'get_current_user') as m:
         m.return_value = self.user
       response = self.fetch(handler.ROUTE, method='GET')
-    self.assertEqual(response.code, 200)
+    self.assertEqual(response.code, 200, msg=handler.__name__)
```
Comment (olivierteboul) [inline icubam/backoffice/server_test.py:51]: nice

File: icubam/backoffice/templates/base.html [hunk 8]
```diff
@@ -141,7 +147,7 @@ <h5>Title</h5>
       Where is the ICU?
     </div>
     <!-- Default to the left -->
-    <strong>Copyright &copy; 2014-2019 <a href="https://adminlte.io">AdminLTE.io</a>.</strong> All rights reserved.
+    <strong>Copyright &copy; 2020 ICUBAM</strong> All rights reserved.
```
Comment (olivierteboul) [inline icubam/backoffice/templates/base.html:150]: good job ;)

**Pull request merged**